std/range/package.d

```diff
@@ -8606,6 +8606,385 @@ if (isForwardRange!Source && hasLength!Source)
+    endIndexes = range specifying the indexes of the elements that will not be included in the current chunk
+
+Returns: Range of chunks of variable lengths
+*/
+auto chunks(Source, Indexes)(Source source, Indexes endIndexes)
+if (isInputRange!Source && isInputRange!Indexes && is(ElementType!Indexes : size_t))
+{
+    struct Chunks(Source, Indexes)
+    {
+        static if (isForwardRange!Source && isForwardRange!Indexes)
+        {
+            this(Source source, Indexes endIndexes, size_t offset = 0, UseMaxLength maxLen=UseMaxLength(false))
+            {
+                import std.algorithm.sorting : isStrictlyMonotonic;
+                assert(endIndexes.isStrictlyMonotonic, "The indexes must be in a strictly increasing order");
+
+                _source = source;
+                _endIndexes = endIndexes;
+                _offset = offset;
+                _maxLen = maxLen;
+                if (!_source.empty)
+                    _isEmpty = false;
+                else
+                    _isEmpty = true;
+            }
+
+            /// Input range primitives. Always present.
+            @property bool empty()
+            {
+                return _isEmpty;
+            }
+
+            /// Ditto
+            @property auto front()
+            {
+                assert(!empty, "Attempting to fetch the front of an empty range");
+
+                if (_endIndexes.empty && !_source.empty)
+                    return _source.save;
+
+                assert(_endIndexes.front != 0, "Attempting to create a chunk of length 0");
+
+                return _source.save.take(_endIndexes.front()-_offset);
+            }
+
+            /// Ditto
+            void popFront()
+            {
+                assert(!empty, "Attempting to popFront an empty range");
+
+                // source range is longer than index
+                if (_endIndexes.empty && !_source.empty)
+                {
+                    _isEmpty = true;
+                    return;
+                }
+
+                _source.popFrontN(_endIndexes.front()-_offset);
+                _offset = _endIndexes.front();
+                _endIndexes.popFront();
+
+                if (_maxLen)
+                {
+                    _maxLen.decrease();
+                    _isEmpty = _maxLen.empty;
+                }
+
+                // length of the source range is smaller than or equal to the index
+                if (_source.empty)
+                    _isEmpty = true;
+            }
+
+            /// Forward range primitives
+            @property typeof(this) save()
+            {
+                return typeof(this)(_source, _endIndexes, _offset);
+            }
+
+            static if (isInfinite!Source && hasLength!Indexes)
+            /// Length is only provided for infinite sources, when hasLength!Indexes is true
+                @property size_t length()
+                {
+                    return _endIndexes.length + 1;
+                }
```

Comment:
length should be provided for `hasLength!Indexes && hasLength!Source && isBidirectionalRange!Indexes`

Reply:
Still not addressed. Would add that the mode has to be `ByIndex`.